src/persist/src/nemesis/generator.rs

```diff
@@ -7,89 +7,689 @@
 };
 
+/// Configuration of the relative probabilities of producing various
+// operations.
+#[derive(Clone, Debug, Serialize)]
+pub struct GeneratorConfig {
+    pub write_unsealed_weight: u32,
+    pub write_sealed_weight: u32,
+    pub seal_weight: u32,
+    pub take_snapshot_weight: u32,
+    pub read_snapshot_weight: u32,
+}
+
+impl GeneratorConfig {
+    fn all_operations() -> GeneratorConfig {
+        GeneratorConfig {
+            write_unsealed_weight: 1,
+            write_sealed_weight: 1,
+            seal_weight: 1,
+            take_snapshot_weight: 1,
+            read_snapshot_weight: 1,
+        }
+    }
+}
+
+impl Default for GeneratorConfig {
+    fn default() -> Self {
+        Self::all_operations()
+    }
+}
+
+struct GeneratorState {
+    sealed_by_stream: HashMap<String, u64>,
+    snap_id: SnapshotId,
+    outstanding_snaps: HashMap<SnapshotId, String>,
+}
+
+impl Default for GeneratorState {
+    fn default() -> Self {
+        GeneratorState {
+            sealed_by_stream: HashMap::new(),
+            snap_id: SnapshotId(0),
+            outstanding_snaps: HashMap::new(),
+        }
+    }
+}
+
+enum ReqGenerator {
+    WriteUnsealed,
+    WriteSealed,
+    Seal,
+    TakeSnapshot,
+    ReadSnapshot,
+}
+
+fn rng_stream(rng: &mut SmallRng) -> String {
+    // TODO: This is meant to be Zipf, use it once rust-random/rand/1136 lands.
+    rng.gen_range('a'..'e').to_string()
+}
+
+fn rng_key(rng: &mut SmallRng) -> String {
+    // TODO: This is meant to be Zipf, use it once rust-random/rand/1136 lands.
+    rng.gen_range('a'..'e').to_string()
+}
+
+impl ReqGenerator {
+    fn write_unsealed(rng: &mut SmallRng, state: &mut GeneratorState) -> Req {
+        let stream = rng_stream(rng);
+        let key = rng_key(rng);
+        let stream_sealed_ts = state
+            .sealed_by_stream
+            .get(&stream)
+            .copied()
+            .unwrap_or_default();
+        let ts = stream_sealed_ts + rng.gen_range(0..5);
+        // TODO: Tables and sources generally will be using -1 and
+        // 1. 0, -2, and 2 might find some more interesting edge
+        // cases. Is it worth expanding this to something with a
+        // wider range (and possibly a non-uniform distribution)
+        // once we start thinking of operator persistence? Perhaps
+        // overflows, what else might be relevant here then?
+        let diff = rng.gen_range(-2..=2);
+        Req::Write(WriteReq {
+            stream,
+            update: ((key, ()), ts, diff),
+        })
+    }
+
+    fn write_sealed(rng: &mut SmallRng, state: &mut GeneratorState) -> Req {
+        let key = rng_key(rng);
+        let (stream, ts) = state
+            .sealed_by_stream
+            .iter()
+            .filter(|(_, ts)| **ts > 0)
+            .choose(rng)
+            .map(|(stream, ts)| (stream.clone(), rng.gen_range(0..*ts)))
+            .unwrap_or_else(|| (rng_stream(rng), 0));
```

Comment:
if you hit this `unwrap_or_else` call then the write will not error right? it seems like hitting this line means that we violated a precondition of calling this function "at least one source needs to be sealed to a ts greater than zero" and instead of the unwrap this should be a nemesis internal error?

Reply:
Yeah, that's exactly right. I was thinking it'd be nice for this to be gracefully handled but seems like the confusion isn't worth it. read_snapshot had something similar, I fixed that one too.